### PR TITLE
review phase 2

### PR DESCRIPTION
**Pull request by gandrewstone**
fix iblt reset/resize issue & add unit test, document iblt and graphene_set APIs, replace double lookups in std containers with a single insert, don't calculate cheaphashes twice

### CODE REVIEW

File: src/graphene.cpp [hunk 1]
```diff
@@ -475,49 +475,45 @@ bool CGrapheneBlock::process(CNode *pfrom,
...
         // Add full transactions included in the block
-        for (auto &tx : vAdditionalTxs)
-        {
-            const uint256 &hash = tx->GetHash();
-            uint64_t cheapHash = hash.GetCheapHash();
-
-            if (mapPartialTxHash.count(cheapHash)) // Check for collisions
-                collision = true;
+        if (!collision)
+            for (auto &tx : vAdditionalTxs)
```
Comment (ptschip) [inline src/graphene.cpp]: just for code readability would be good to see a parenthesis  here

+if (!collision)
+{
...
+}

File: src/graphene.cpp [hunk 2]
```diff
@@ -475,49 +475,45 @@ bool CGrapheneBlock::process(CNode *pfrom,
...
         {
             uint64_t cheapHash = kv.first.GetCheapHash();
-
-            if (mapPartialTxHash.count(cheapHash)) // Check for collisions
-                collision = true;
-
-            mapPartialTxHash[cheapHash] = kv.first;
+            auto ir = mapPartialTxHash.insert(std::make_pair(cheapHash, kv.first));
+            if (!ir.second)
+                collision = true; // insert returns false if no insertion
```
Comment (ptschip) [inline src/graphene.cpp:480]: I like this approach but shouldn't this be if (!ir) and not ir.second?  also in a couple of other places.

File: src/graphene.cpp [hunk 3]
```diff
@@ -475,49 +475,45 @@ bool CGrapheneBlock::process(CNode *pfrom,
...
         // We don't have to keep the lock on mempool.cs here to do mempool.queryHashes
         // but we take the lock anyway so we don't have to re-lock again later.
         LOCK(cs_xval);
-        mempool.queryHashes(memPoolHashes);
-
-        for (const uint256 &hash : memPoolHashes)
+        if (!collision)
         {
-            uint64_t cheapHash = hash.GetCheapHash();
+            mempool.queryHashes(memPoolHashes);
```
Comment (ptschip) [inline src/graphene.cpp:488]: why the extra parentheses here?

File: src/graphene.cpp [hunk 4]
```diff
@@ -475,49 +475,45 @@ bool CGrapheneBlock::process(CNode *pfrom,
...
         if (!collision)
         {
-            std::vector<uint256> localHashes;
-            for (const std::pair<uint64_t, uint256> &kv : mapPartialTxHash)
-                localHashes.push_back(kv.second);
-
             try
             {
-                std::vector<uint64_t> blockCheapHashes = pGrapheneSet->Reconcile(localHashes);
+                std::vector<uint64_t> blockCheapHashes = pGrapheneSet->Reconcile(mapPartialTxHash);
```
Comment (ptschip) [inline src/graphene.cpp:519]: good, i missed that one.

File: src/graphene.cpp (hunk 2 above)
Comment (gandrewstone) [reply src/graphene.cpp:480]: https://en.cppreference.com/w/cpp/container/map/insert
overload 2: template< class P > std::pair<iterator,bool> insert( P&& value );

File: src/graphene.cpp (hunk 3 above)
Comment (gandrewstone) [reply src/graphene.cpp:488]: I want to skip both the mempool.queryHashes(memPoolHashes) and the for loop if there is already a collision

**Comment by ptschip** [comment]
ah, ok good

File: src/graphene.cpp (hunk 3 above)
Comment (ptschip) [reply src/graphene.cpp:488]: yup ok, i didn't see the !collision...

**Review by ptschip — APPROVED** [review]
ok, looks all good to me...

File: src/iblt.h [hunk 5]
```diff
@@ -69,12 +69,18 @@ class HashTableEntry
 class CIblt
 {
 public:
+    // Default constructor builds a 0 size IBLT, so is meant for two-phase construction.  Call resize() before use
     CIblt();
+    // Pass the expected number of entries in the IBLT table.  If the number of entries exceeds
+    // the expected, the false positive rate will rise
```
Comment (bissias) [inline src/iblt.h]: Small thing; technically it's the decode failure rate that will increase. So I would rephrase as follows

`// Pass the expected number of entries in the IBLT table.  If the number of entries exceeds`
`// the expected, then the decode failure rate will increase dramatically.`
